src/lib.rs

```diff
@@ -25,7 +139,50 @@ mod impl_details {
 
 #[cfg(feature = "gecko-ffi")]
 mod impl_details {
+    // Support for briding a gecko nsTArray verbatim into a ThinVec.
+    //
+    // ThinVec can't see copy/move/delete implementations
+    // from C++
+    //
+    // The actual layout of an nsTArray is:
+    //
+    // ```cpp
+    // struct {
+    //   uint32_t mLength;
+    //   uint32_t mCapacity: 31;
+    //   uint32_t mIsAutoArray: 1;
+    // }
+    // ```
+    //
+    // Rust doesn't natively support bit-fields, so we manually mask
+    // and shift the bit. When the "auto" bit is set, the header and buffer
+    // are actually on the stack, meaning the ThinVec pointer-to-header
+    // is essentially an "owned borrow", and therefore dangerous to handle.
+    // There are no safety guards for this situation.
+    //
+    // On little-endian platforms, the auto bit will be the high-bit of
+    // our capacity u32. On big-endian platforms, it will be the low bit.
+    // Hence we need some platform-specific CFGs for the necessary masking/shifting.
+    // 
+    // ThinVec won't ever construct an auto array. They only happen when
+    // bridging from C++. This means we don't need to ever set/preserve the bit.
+    // We just need to be able to read and handle it if it happens to be there.
+    //
+    // Handling the auto bit mostly just means not freeing/reallocating the buffer.
+    
+    // A Mask for loading the auto bit
+    #[cfg(target_endian = "little")]
     pub const AUTO_MASK: u32 = 1 << 31;
+    #[cfg(target_endian = "big")]
+    pub const AUTO_MASK: u32 = 1;
+
+    // How much of a shift to perform to load/store the capacity
+    #[cfg(target_endian = "little")]
+    pub const CAP_SHIFT: u32 = 0;
+    #[cfg(target_endian = "big")]
+    pub const CAP_SHIFT: u32 = 1;
```

Comment:
It seems a bit like we should perhaps stop using bitfields on the c++ side, so that we can be more confident here. The layout of bitfields I believe is implementation defined, so it's possible these checks aren't fully portable.